http3/server.go

```diff
@@ -497,19 +538,51 @@ func (s *Server) handleConn(serverCtx context.Context, conn quic.Connection) err
 	// Process all requests immediately.
 	// It's the client's responsibility to decide which requests are eligible for 0-RTT.
 	for {
-		str, datagrams, err := hconn.acceptStream(serverCtx)
+		str, datagrams, err := hconn.acceptStream(s.graceCtx)
 		if err != nil {
-			// close the connection if the server was closed
-			if errors.Is(err, context.Canceled) {
+			// not gracefully closed, close the connection immediately
+			if s.closeCtx.Err() != nil {
 				conn.CloseWithError(quic.ApplicationErrorCode(ErrCodeNoError), "")
+				return http.ErrServerClosed
+			}
+
+			// gracefully closed, send GOAWAY frame and wait for requests to complete or grace period to end
+			// new requests will be rejected and shouldn't be sent
+			if s.graceCtx.Err() != nil {
+				b = (&goawayFrame{
+					StreamID: lastID + 4,
+				}).Append(b[:0])
+				// set a deadline to send the GOAWAY frame
+				ctrlStr.SetWriteDeadline(time.Now().Add(goawayTimeout))
+				ctrlStr.Write(b)
+
+				select {
+				// some requests may be still running, wait for them to finish before returning
+				// do nothing let quic to deliver the data the peer
+				case <-handlingDoneChan:
```

Comment:
We can't return early here. We still need to close the connection if the client handle the GOAWAY frame.

Reply:
> @marten-seemann There is some race going on. If I close the connection immediately like you did. `graceful shutdown successfully with no timeout` will fail because although the request handling is done. The data written isn't delivered yet to the peer (I think, from `SendStream` documentation, `CancelWrite` specifically), causing the response to fail. How do I deal with this situation? That the data is delivered to the peer. Do I need to use the channel in the client in the tests to simulate this situation?

That will cause this problem.

Reply:
It doesn't. Just don't close the connection before cancelation of the context.

Reply:
If client side reading can be changed to read the available data first, and ignore the connection error because the data is already there and fully read. Then it can be changed.

Reply:
That's not how QUIC works. But as discussed in https://github.com/quic-go/quic-go/pull/4669#issuecomment-2403747650, it's the client who closes the connection.

Reply:
Do you mean the context of the quic.Connection? So the server will close the connection when either it's forcibly shut down or when the client closes?

Reply:
We don't need any context here. We just need to close the connect when the close context is cancelled.

Reply:
Just wait for the `closeCtx` to be done, then?